### PR TITLE
added Position.equals, string.split

### PR DESCRIPTION
**Pull request by Choumiko**
string.split splits a string into a table by a seperator (with optional lua pattern support), the spec should serve as documentation/use cases

### CODE REVIEW

**Comment by credomane** [comment]
Could you shove string.is_empty in this PR too? I use this one a lot. Just haven't made a PR for it yet.
```lua
function string.is_empty(s)
  return s == nil or s == ''
end
```

File: stdlib/area/position.lua [hunk 1]
```diff
@@ -132,6 +132,14 @@ function Position.manhattan_distance(pos1, pos2)
     return math.abs(pos2.x - pos1.x) + math.abs(pos2.y - pos1.y)
 end
 
+--- Whether 2 positions are equal
+-- @param pos1 the first position
+-- @param pos2 the second position
+-- @return true if positions are equal
+function Position.equals(pos1, pos2)
+    return (Position.distance_squared(pos1,pos2) == 0)
```
Comment (Afforess) [inline stdlib/area/position.lua]: This is a really expensive way to test it. Maybe add a line to test if pos1 ~= nil and pos1 == pos2, then do the distance_squared test.

**Comment by Afforess** [comment]
Looks pretty good. Concerned the equality check is expensive, but I'm open to counterarguments.

**Comment by credomane** [comment]
This is my take on it.
```lua
function Position.equals(pos1, pos2)
    pos1 = Position.to_table(pos1)
    pos2 = Position.to_table(pos2)
    return pos1.x == pos2.x and pos1.y == pos2.y
end
```

**Comment by Afforess** [comment]
@credomane comparing floats directly may not be entirely safe, mostly due to the IEEE floating point representation insanity. Probably only works for smaller positions.

**Comment by Afforess** [comment]
Merged and used @credomane's equality check. Nice work @Choumiko!

**Comment by credomane** [comment]
Knew it was too easy. I always forget about nature of floating point and the resulting inaccuracies.  Everything else I come up with is at least as expensive as the current solution if not more expensive.

I got nothing that would be of real help. I threw two other versions of my crappy attempts to find a better way for giggles. Anything else I came up was even worse than these. So by all means laugh at me for these two. :D

```lua
function Position.equals(pos1, pos2)
    pos1 = Position.to_table(pos1)
    pos2 = Position.to_table(pos2)
    return math.floor(pos1.x*1000) == math.floor(pos2.x*1000) and math.floor(pos1.y*1000) == math.floor(pos2.y*10000)
end

function Position.equals(pos1, pos2)
    pos1 = Position.to_table(pos1)
    pos2 = Position.to_table(pos2)
    local minx = math.min(pos1.x, pos2.x)
    local maxx = math.max(pos1.x, pos2.x)
    local miny = math.min(pos1.y, pos2.y)
    local maxy = math.max(pos1.y, pos2.y)

    return (maxx - minx) < 0.0002 and (maxy - miny) < 0.0002
end
```

**Comment by OvermindDL1** [comment]
When it comes to floating point they should never be directly compared.  The second Position.equals in @credomane post is the right way, although 0.0002 is very arbitrary and may not work well in all cases.  Specifically a scaled epsilon should be used, which should be about 1.19e-7f for comparing points between -1.0 to 1.0, which will not work well in this case, which gets into the more hairy epsilon scaling code (because every time the float scaler is doubled then the epsilon needs to be doubled, but only in those jumps, not linearly).  0.0002 should be fine for most cases, but it should probably be passed in via an optional parameter that defaults to 0.0002 so it can be changed by modders in proper cases.

Floating point sucks

P.S. Does the Position.to_table should really need to be done?  Perhaps better to decompose the pos1/pos2 into 4 `local` variables of `pos1_x`, `pos1_y`, etc... since this is very likely to be code that is called in a tight loop?

**Comment by credomane** [comment]
> although 0.0002 is very arbitrary and may not work well in all cases

Yeah...I actually meant for it to be a 1 not a 2 but didn't bother correcting since I already posted when I saw it. That is in there for "good enough for government work" precision error correction.  Should be functionally the same as `math.floor((maxx - minx)*1000) < 1` only faster unless I'm way off on just how expensive floating point comparison really is.

> P.S. Does the Position.to_table should really need to be done? 

Yes. These are both valid and the same positions in Factorio `{1, 2}` and `{x=1, y=2}` 
Position.to_table enforces the position to always be in the form `{x=1, y=2}`.

> Oh, also the maxx-minx and same for y need either an abs done or a `<epsilon and >-epsilon` done.

The min and max ensure that extra step isn't needed. As far as my napkin math says anyways.

    1 - 1 = 0 (return true)
    1 - 0 = 1 (return false)
    1 - -1 = 2 (return false)
    0 - 1 = -1 (error min() and max() borked. one is greater than zero)
    0 - 0 = 0 (return true)
    0 - -1 = 1 (return false)
    -1 - 1 = -2 (error min() and max() borked. one is greater than negative one)
    -1 - 0 = -1 (error min() and max() borked. zero is greater than negative one)
    -1 - -1 = 0 (return true)

**Comment by OvermindDL1** [comment]
The extra step would be faster than a function call that does the extra comparison anyway though?  :-)

**Comment by Afforess** [comment]
@Choumiko to fix https://github.com/Afforess/Factorio-Stdlib/issues/34 I had to tweak how string.split handled the pattern argument case. Let me know what you think.

**Comment by Afforess** [comment]
@OvermindDL1 @credomane 

Hopefully this revamped Position.equals is satisfactory to everyone:

```
-- see: https://en.wikipedia.org/wiki/Machine_epsilon
Position._epsilon = 1.19e-07

--- Whether 2 positions are equal
-- @param pos1 the first position
-- @param pos2 the second position
-- @return true if positions are equal
function Position.equals(pos1, pos2)
    -- optimize for a shallow equality check first
    if pos1 == pos2 then return true end

    local epsilon = Position._epsilon
    local abs = math.abs
    if #pos1 == 2 and #pos2 == 2 then
        return abs(pos1[1] - pos2[1]) < epsilon and abs(pos1[2] - pos2[2]) < epsilon
    elseif #pos1 == 2 and #pos2 == 0 then
        return abs(pos1[1] - pos2.x) < epsilon and abs(pos1[2] - pos2.y) < epsilon
    elseif #pos1 == 0 and #pos2 == 2 then
        return abs(pos1.x - pos2[1]) < epsilon and abs(pos1.y - pos2[2]) < epsilon
    elseif #pos1 == 0 and #pos2 == 0 then
        return abs(pos1.x - pos2.x) < epsilon and abs(pos1.y - pos2.y) < epsilon
    end
end
```

**Comment by credomane** [comment]
Ah. So that is what @OvermindDL1 was talking about with abs and epsilon. I had no idea epsilon was a thing. Originally thought it was a misspelling as was trying to figure out how abs would be of any use. I was missing the other part of the math magic trick. Should be far more reliable than my quick-n-dirty compare would ever be. lol

While we are at it and just because should we add a Position.fuzzy_equals (better name needed) that considers two positions on the same tile as being equal? Would just math.floor each of the pos x/y before comparing work?

````lua
--- Whether 2 positions are on the same tile
-- @param pos1 the first position
-- @param pos2 the second position
-- @return true if positions are on same tile
function Position.fuzzy_equals(pos1, pos2)
    if not pos1 or not pos2 then return false end
    -- optimize for a shallow equality check first
    if pos1 == pos2 then return true end

    local epsilon = Position._epsilon
    local abs = math.abs
    local floor = math.floor
    if #pos1 == 2 and #pos2 == 2 then
        return abs(floor(pos1[1]) - floor(pos2[1])) < epsilon and abs(floor(pos1[2]) - floor(pos2[2])) < epsilon
    elseif #pos1 == 2 and #pos2 == 0 then
        return abs(floor(pos1[1]) - floor(pos2.x)) < epsilon and abs(floor(pos1[2]) - floor(pos2.y)) < epsilon
    elseif #pos1 == 0 and #pos2 == 2 then
        return abs(floor(pos1.x) - floor(pos2[1])) < epsilon and abs(floor(pos1.y) - floor(pos2[2])) < epsilon
    elseif #pos1 == 0 and #pos2 == 0 then
        return abs(floor(pos1.x) - floor(pos2.x)) < epsilon and abs(floor(pos1.y) - floor(pos2.y)) < epsilon
    end

    return false
end
````

**Comment by Afforess** [comment]
Not against a "fuzzy equals" but it then belongs in Tile module and can
just be Tile.equals.

**Comment by Afforess** [comment]
@credomane 

Next cold and dreary evening, grab a warm drink and settle down with ["What every computer scientist should know about floating-point arithmatic"](https://ece.uwaterloo.ca/~dwharder/NumericalAnalysis/02Numerics/Double/paper.pdf).

**Comment by credomane** [comment]
> Tile module and can just be Tile.equals.

That works.

>Next cold and dreary evening, grab a warm drink and settle down with "[What every computer scientist should know about floating-point arithmatic](https://ece.uwaterloo.ca/%7Edwharder/NumericalAnalysis/02Numerics/Double/paper.pdf)".

I glanced at that. Oh goody it is a long one. :D

**Comment by OvermindDL1** [comment]
> I glanced at that. Oh goody it is a long one. :D
Lol, and that seems like a decent article to de-code the IEEE floating point spec.  ;-)

Yes, floating point is horrible, I try not to use it in games, fixed-point is usually better, integers at a scale you need even more so.  Sadly floating point is really needed with a lot of math for speed reasons.

**Comment by credomane** [comment]
>  Sadly floating point is really needed with a lot of math for speed reasons

Really? Interesting. I always thought floating point operations were expensive and slow compared to integer operations. Learning all kinds of stuff with FP today.

**Comment by OvermindDL1** [comment]
That was true back in the day, but modern CPU's (by modern I mean modern for me, this has been true for probably 20 years worth of CPU's now, at least 15) are exceptional at floating point operations in a variety of ways.  Sure a direct operation of simple operations like multiplication of integers and multiplication of floats the integers would be faster, however when you start doing multiple operations, say a multiplication and an addition together the CPU can recognize that pattern and take an optimized path that do those together.  Almost all floating point operations are done in either vectorized sets, which can be pipelined along SSE-style paths, or can become complex calls to perform multiple operations at once.  It is actually quite a fascinating study.  Do note this is mostly true only on x86/x64, arch's like Arm have significantly reduced floating point processing abilities and you will rarely if ever see a speed boost there.

If you know how x86/x64 works on hardware, and your structure your code just right, floating point can almost always outperform integers, the trade-off is of course the performance loss, but if you understand how and where that happens then you can minimize it, but no matter how well you minimize it, if you deal with it often it will always seem to some how, some way, in some time, bite you in the butt.
